packages/xstate-wallet/src/workflows/challenge-channel.ts

```diff
@@ -0,0 +1,176 @@
+
+interface Schema extends StateSchema<Context> {
+  states: {
+    init: {};
+    waitForResponseOrTimeout: {};
+    submitTransaction: {};
+    retry: {};
+    done: {};
+    waitMining: {};
+    failure: {};
+  };
+}
+
+export type WorkflowState = State<Context, Event, Schema, Typestate>;
+
+export type StateValue = keyof Schema['states'];
+
+interface ChainEvent {
+  type: 'CHAIN_EVENT';
+  turnNumRecord: BigNumber;
+  finalizesAt: BigNumber;
+  finalized: boolean;
+}
+
+const noChallengeOnchain: Guard<Initial, ChainEvent> = {
+  type: 'xstate.guard',
+  name: 'noChallengeOnchain',
+  predicate: (context, {finalizesAt}) => finalizesAt.lte(Zero)
+};
+
+// const someOtherChallengeOnchain: Guard<Initial, ChainEvent> = {
+//   type: 'xstate.guard',
+//   name: 'myTurnNow',
+//   predicate: (context, event) => false // TODO: Add challenge state to context
+// };
+
+const challengeOnchainAsExpected: Guard<Initial, ChainEvent> = {
+  type: 'xstate.guard',
+  name: 'challengeOnchainAsExpected',
+  predicate: (context, {finalizesAt, finalized}) => finalizesAt.gt(0) && !finalized
+};
+
+const challengeFinalized: Guard<Initial, ChainEvent> = {
+  type: 'xstate.guard',
+  name: 'challengeFinalized',
+  predicate: (context, {finalized}) => finalized
+};
+
+const submitChallengeTransaction = (store: Store) => async ({channelId}: Initial) => {
+  const {support, myAddress} = await store.getEntry(channelId);
+  const privateKey = await store.getPrivateKey(myAddress);
+  return await store.chain.challenge(support, privateKey);
+};
+
+const observeOnChainChannelStorage = (store: Store, channelId: string) =>
+  store.chain.chainUpdatedFeed(channelId).pipe(
+    map<ChannelChainInfo, ChainEvent>(({finalized, channelStorage}) => ({
+      type: 'CHAIN_EVENT',
+      finalized,
+      ...channelStorage
+    }))
+  );
+
+const setTransactionId = assign<Context, DoneInvokeEvent<string>>({
+  transactionId: (context, {data}) => data
+});
+
+export const machine = (
+  store: Store,
+  context: Initial
+): StateMachine<Context, Schema, Event, Typestate> =>
+  createMachine<Context, Event, Typestate>({
+    context,
+
+    strict: true,
+    id: 'challenge-channel',
+    initial: 'init',
+
+    on: {
+      CHAIN_EVENT: [
+        {
+          target: 'waitForResponseOrTimeout',
+          cond: challengeOnchainAsExpected
+        },
+        {
+          target: 'done',
+          cond: challengeFinalized
+        }
+      ]
+    },
+
+    states: {
+      init: {
+        // TODO: Figure out how to make invoke work at root-level here. Seems to cause
+        // an infinite loop if entry is on the root-level of the machine.
+        entry: assign<any>({
+          chainWatcher: ({channelId}) => spawn(observeOnChainChannelStorage(store, channelId))
+        }),
+
+        on: {
+          CHAIN_EVENT: {
+            target: 'submitTransaction',
+            cond: noChallengeOnchain
+          }
+        }
+      },
+
+      waitForResponseOrTimeout: {},
```

Comment:
Might be worth moving this below `waitMining`, as this is where it naturally comes in the workflow?